dspace-uclouvain/src/main/java/org/dspace/uclouvain/pdfAttestationGenerator/xmlworkflow/actions/SendEmailAttestationAction.java

```diff
@@ -0,0 +1,270 @@
+            return new ActionResult(ActionResult.TYPE.TYPE_ERROR);
+        }
+        return new ActionResult(ActionResult.TYPE.TYPE_OUTCOME, ActionResult.OUTCOME_COMPLETE);
+    }
+
+    @Override
+    public List<String> getOptions() {
+        return new ArrayList<String>();
+    }
+
+    /**
+     * Send an email to all the submitters with the PDF attestation attached to it.
+     * @param metadata: A HashMap containing all the metadata of the submission.
+     * @param dspaceItem: The DSpace item corresponding to the submission.
+     * @param templatePath: The path to the email template to use.
+     * @param attestation: The PDF attestation to attach to the email.
+     * @throws Exception
+     */
+    private void sendSubmitterEmail(HashMap<String, List<String>> metadata, Item dspaceItem, String templatePath, InputStream attestation) throws Exception {
+        Email email = this.generateBaseEmail(metadata, templatePath, metadata.get("authors_email"), attestation);
+        email.send();
+    }
+
+    /**
+     * Send an email to all the promoters with the PDF attestation attached to it.
+     * @param metadata: A HashMap containing all the metadata of the submission.
+     * @param dspaceItem: The DSpace item corresponding to the submission.
+     * @param templatePath: The path to the email template to use.
+     * @param attestation: The PDF attestation to attach to the email.
+     * @throws Exception
+     */
+    private void sendPromoterEmail(HashMap<String, List<String>> metadata, Item dspaceItem, String templatePath, InputStream attestation) throws Exception {
+        Email email = this.generateBaseEmail(metadata, templatePath, metadata.get("advisors_email"), attestation);
+        this.appendUrlsToEmail(metadata, email, dspaceItem);
+        email.send(); 
+    }
+
+    /**
+     * Send an error email to both the submitters and the promoters with the stacktrace of the exception.
+     * This is done when an error is occurring when generating the PDF attestation (or a mail).
+     * @param metadata: A HashMap containing all the metadata of the submission.
+     * @param submitter: The EPerson who submitted the item.
+     * @param dspaceItem: The DSpace item corresponding to the submission.
+     * @param e: The exception that occurred.
+     * @throws Exception
+     */
+    private void sendErrorEmail(HashMap<String, List<String>> metadata, EPerson submitter, Item dspaceItem, Exception e) throws Exception {
+        Email email = Email.getEmail(this.source + "/config/emails/pdf_attestation_error");
+        // TODO: Use given emails
+        // for (String address : metadata.get("authors_email")) {
+        //     email.addRecipient(address);
+        // }
+        // for (String address : metadata.get("advisors_email")) {
+        //     email.addRecipient(address);
+        // }
+        email.addRecipient("michael.pourbaix@uclouvain.be");
```

Comment:
Is the classic config could be hacked if a config key is available into "local.cfg".
It will prevent a "hard" code modification when passing app in production

```java
String[] recipients = this.getAttestationRecipients();
if (recipients.size == 0) {
    throw new Exception("no recipient bla bla...");
}
for (String address; recipients) {
    email.addRecipient(address);
}
...


private Iterator<String> getAttestationRecipients() {
    recipients = configService.getKey("override_attestation_recipients");
    if (recipients.size > 0 ) {
         return recipients;
    }else{
        List<String> emails = new ArrayList<>();
        for (String fieldName: ['authors_email', 'advisors_email']) {
            for (String address : metadata.get("authors_email")) {
                email.append(address);
            }
         }
         return emails;
    }
}
```